docs/rims.md

```diff
@@ -0,0 +1,203 @@
+* Signature - digital signature of the RIM
+
+Example Base RIM:
+
+```
+Software Identity Name: Example.com BIOS
+  version : 01 (MUST BE BIOS version)
+  tagId: 94f6b457-9ac9-4d35-9b3f-78804173b65as (from TPM event log)
+  tagVersion:0
+Entity (tagCreator) Name
+  Regid: http://Example.com
+  Role: softwareCreator tagCreator
+Links:
+  installation media url: https://Example.com/support/ProductA/firmware/installfiles
+Meta:
+  colloquialVersion: Firmware_2019
+  Edition: 12
+  Product: ProductA
+  Revision: r2
+  PayloadType: Indirect
+  PlatformManufacturerStr: Example.com
+  PlatformManufacturerId: 00201234
+  PlatformModel: ProductA
+  PlatformVersion:01
+  FirmwareManufacturerStr: BIOSVendorA
+  FirmwareManufacturerId: 00213022
+  FirmwareModel:A0
+  FirmwareVersion: 12
+  BindingSpec: PC Client RIM
+  BindingSpecVersion: 1.2
+  pcURIlocal: /boot/tcg/manifest/swidtag
+Payload:
+  Directory: /boot/tcg/manifest/rim
+    File1: Example.com.Notebook.3.rimel
+     Version: 01.00
+     size= 15400
+     SHA256 hash: a314fc2dc663ae7a6b6bc6787594057396e6b3f569cd50fd5ddb4d1bbafd2b6a
+     supportRIMFormat: TPM Event Log Assertions
+    File2: Example.com.Notebook.3.rimpcr
+     Version: 01.00
+     size: 1024
+     SHA256:hash: 532eaabd9574880dbf76b9b8cc00832c20a6ec113d682299550d7a6e0f345e25
+     supportRIMFormat: TPM PCR Assertions
+Signature: (Enveloped) 
+  sigAlgorithm: rsa-sha256 (MUST be a TCG listed algorithm)
+  hashAlgorithm: sha256 (MUST be a TCG listed algorithm)
+  keyInfo: subjectKeyIdentifier OR X509Data
+  digest: 8A4C98225D855D388B42FD8AA4FA1646CFAC666E43FD6A7B950BB3E0B12134AA
+  signature <hex>
+Table 1 RIM IM Element-Attribute table
+```
+
+The RIM above contains two files, one with TPM event log, second with reference
+PCRs. These Support RIM formats are called TPM Event Log Assertions and TPM PCR
+Assertions. These formats are mandated by PC Client RIM binding specification
+(in other words the TCG PC Client Reference Integrity Manifest Specification).
+
+The content of the files is simple:
+
+* TPM Event Log Assertions - raw TPM event log dump
+* TPM PCR Assertions - raw output from TPM2_PCR_Read command (without response
+  header):
+    * `UINT32 pcrUpdateCounter`
+    * `TPML_PCR_SELECTION pcrSelectionOut`
+    * `TPML_DIGEST pcrValues`
+
+There may be multiple Support RIMs with PCRs or Event Logs.
+
+There is a tool abel to generate the RIMs in
+[Host Integrity at Runtime and Start-up (HIRS) repository](https://github.com/nsacyber/HIRS/tree/master/tools/tcg_rim_tool)
+
+## RIM packaging
+
+The RIM format is not processor-friendly. The data is not structured. RIM can
+be well described by XML, but in such form it is almost not consumable by a
+small MCU. Thus a few modifications must be made in order to transfer such RIM
+and its payload to fobnail. The following structures are proposed (as close to
+specification as possible)to be used by fobnail firmware (all strings are NULL
+terminated unless stated otherwise):
```

Comment:
@miczyg1 Can't we use CBOR for this? We are using it already for something else.

Reply:
True, the RIM specification also defines CBOR notation/representation of the RIMs. However I can see some descriptions are in draft status, e.g. https://www.ietf.org/archive/id/draft-ietf-sacm-coswid-20.html and https://datatracker.ietf.org/doc/html/draft-birkholz-rats-coswid-rim-02
They constantly are being expired.

I would have to go through these specs and change the document. However, I cannot see any advantage of using the CBOR over my proposal. You would have to store most of the strings anyway.

Reply:
@arturkow2000 what's left to close this one?